teps/0052-tekton-results-automated-run-resource-cleanup.md

```diff
@@ -0,0 +1,213 @@
+this creates a problem for the Pipeline controller - as the build history grows
+over time, so does the corpus of data the controller needs to handle in its data
+plane. Over time this grows unbounded, significantly degrading the performance
+of the cluster as more Runs are completed.
+
+Most of this data is unnecessary for the Pipeline controller to hold on to -
+once a Run is complete it no longer needs to be stored in etcd since the
+Pipeline controller will take no further action on it. With Tekton Results, we
+now have a place to store long term results independent of the Pipeline
+controller, even if the data is removed from the etcd data plane.
+
+Many users have implemented custom cleanup mechanisms to address this
+([including our own dogfooding cluster](https://github.com/tektoncd/plumbing/blob/2c1808d75b38d6fcaae0a37f5a9dda50545af4c8/tekton/resources/cd/cleanup-template.yaml)).
+This proposal aims to define a supported mechanism for this behavior in
+conjunction with our existing Results storage.
+
+Relevant Issues:
+
+- [tektoncd/pipeline#2332](https://github.com/tektoncd/pipeline/issues/2332)
+- [tektoncd/pipeline#1486](https://github.com/tektoncd/pipeline/pull/1486)
+- [tektoncd/pipeline#1429](https://github.com/tektoncd/pipeline/pull/1429)
+- [tektoncd/experimental#479](https://github.com/tektoncd/experimental/issues/479)
+- [tektoncd/pipeline#2452](https://github.com/tektoncd/pipeline/issues/2452)
+
+### Goals
+
+- Define a mechanism for cleaning up completed Runs to minimize resources the
+  Pipeline controller is responsible for.
+- Preserve Run metadata so that users can continue to query it beyond the
+  lifetime of the Pipeline controller.
+
+### Non-Goals
+
+- Providing a programmatic mechanism for end users to dynamically control deletion behavior.
+
+### Use Cases (optional)
+
+This is a commonly requested feature due to the strain Run history this puts on
+etcd over time. See [Relevant Issues under Motivation](#motivation).
+
+## Requirements
+
+- Run metadata should be available even after deletion from the Pipeline
+  controller data plane.
+- Runs should be deleted as soon as it can be done safely (i.e. we have
+  confirmed the data is stored durably elsewhere).
+- Deletion should occur with some configurable delay to allow other integrations
+  watching Runs to respond to events.
+- Operators should be able to opt in / out to deletion behavior.
+
+## Proposal
+
+### Delete Runs when complete
+
+The
+[Tekton Results watcher](https://github.com/tektoncd/results/blob/main/docs/watcher/README.md)
+already watches for changes to TaskRun and PipelineRuns to upload Records for
+long term storage. Once we have verified that 1)
+[the Run has completed](https://github.com/tektoncd/pipeline/blob/master/docs/runs.md#monitoring-execution-status)
+and 2) the Run Record has been successfully stored, we know we can delete the
+Run without affecting execution.
```

Comment:
how will this actually work? will the controller re-queue the runs to be re-reconciled after the grace period?